### ISSUE TITLE
[refactor] Land or close refactor/importance-sklearn

### ISSUE THREAD

**Issue by thec0dewriter** [thec0dewriter/xeries#15]
**Source**: explicit hand-off in [`specs/003-conditional-permutation-importance/spec.md`](https://github.com/xeries-labs/xeries/blob/feat/sdd-adoption/specs/003-conditional-permutation-importance/spec.md) §"Out of scope / future work":

> The currently parked branch `refactor/importance-sklearn` proposes to delegate the unconditional-PFI path to `sklearn.inspection.permutation_importance` (the conditional path stays hand-rolled because sklearn has no equivalent). When that lands, this Backfilled spec MUST be updated — it is the canonical surface contract.

## Current state

During SDD adoption, the branch `refactor/importance-sklearn` was kept locally only — it was **deliberately not pushed** alongside the SDD bootstrap on `feat/sdd-adoption` so the two reviews stay independent. Tip: `5070839 refactor(importance): delegate unconditional PFI to sklearn`.

## What needs to happen

Decide one of:

### Option A — Land it

1. Push `refactor/importance-sklearn` to `origin`.
2. Rebase on `main`.
3. Confirm the cPFI test suite still passes (`uv run pytest tests/unit/test_permutation.py`) and that conditional behaviour is unchanged on a small but representative dataset.
4. Open a PR. The Constitution Check on the plan side is light — this is a mechanical refactor that **preserves the public surface**, so per the [`specs/README.md`](https://github.com/xeries-labs/xeries/blob/feat/sdd-adoption/specs/README.md) policy it does not need a new spec, but the PR description MUST cite spec 003.
5. As part of the merge, update `specs/003-conditional-permutation-importance/spec.md` FR-003.2 to mention the delegation explicitly (one-line note that the unconditional path now wraps `sklearn.inspection.permutation_importance` while the conditional path remains hand-rolled).

### Option B — Close as `not_planned`

1. Close this issue with `state_reason=not_planned` and a short note on why the refactor was rejected.
2. Delete the local `refactor/importance-sklearn` branch.
3. Update spec 003's "Out of scope / future work" section to remove the dangling pointer.

## Definition of done

This issue is closed only after **either** the PR has merged to `main` and the spec 003 update is in, **or** the branch is deleted and the spec 003 section is cleaned up. The current limbo — a parked local branch referenced by a published spec — is what this ticket exists to resolve.

**Comment by thec0dewriter** [comment]
> **Decision plan for review** — this issue is a *land-or-close* call on a parked branch, not a new feature. There is therefore no `specs/NNN-*/spec.md` to author; the binding contract already lives in [`specs/003-conditional-permutation-importance/spec.md`](https://github.com/xeries-labs/xeries/blob/main/specs/003-conditional-permutation-importance/spec.md). This comment formalises the decision criteria and the mechanical path for whichever branch (Land or Close) is chosen.

# Decision: `refactor/importance-sklearn`

**Branch**: `refactor/importance-sklearn` @ `5070839` (`refactor(importance): delegate unconditional PFI to sklearn`) — currently local-only, deliberately not pushed alongside the SDD bootstrap on `feat/sdd-adoption`.  
**Home repo**: xeries  
**Created**: 2026-04-28

## Context

[`ConditionalPermutationImportance`](https://github.com/xeries-labs/xeries/blob/main/src/xeries/importance/permutation.py) currently hand-rolls **both** the unconditional PFI path (`strategy != "auto"` and no partitioner active) and the conditional path (subgroup-restricted permutation). Spec 003 §"Out of scope / future work" notes:

> The currently parked branch `refactor/importance-sklearn` proposes to delegate the unconditional-PFI path to `sklearn.inspection.permutation_importance` (the conditional path stays hand-rolled because sklearn has no equivalent). When that lands, this Backfilled spec MUST be updated — it is the canonical surface contract.

The parked commit is a **mechanical refactor**: it preserves the `ConditionalPermutationImportance` public surface (constructor signature in spec 003 FR-003.1, `explain` return type in FR-003.4, and the `n_repeats`/`n_jobs`/`random_state` semantics) and only swaps the internal unconditional loop for a `sklearn.inspection.permutation_importance` call.

## Decision criteria

Pick **Option A** (land) if all four hold:

1. The cPFI conditional path test suite (`tests/unit/test_permutation.py`) passes unchanged after rebasing onto `main`.
2. Determinism under fixed `random_state` is preserved (sklearn's seed plumbing is already `Generator`-compatible).
3. The diff against `main` does not change `FeatureImportanceResult.{importances, std, baseline_score, permuted_scores}` field semantics.
4. The added sklearn dependency is **already** in `pyproject.toml` (it is — sklearn is runtime).

Pick **Option B** (close) if any of the four fails or if the maintainers consciously prefer a fully hand-rolled implementation for review-surface reasons.

## Option A — Land it

### Mechanical steps

```bash
# 1. Push the parked branch to origin
git push -u origin refactor/importance-sklearn

# 2. Rebase on main
git fetch origin
git rebase origin/main refactor/importance-sklearn

# 3. Verify the cPFI suite (conditional path is the one we care about)
uv run pytest tests/unit/test_permutation.py -q
uv run pytest tests/integration/ -q -k "permutation"

# 4. Verify determinism
uv run pytest tests/unit/test_permutation.py::test_reproducibility_under_fixed_seed -q  # add if missing

# 5. Verify lint + types
uv run ruff check src tests
uv run ty check src

# 6. Open PR against main, citing spec 003
```

### Spec 003 update (mandatory part of the merge)

FR-003.2 currently reads (paraphrased): *"Permutation is performed by shuffling feature columns and re-scoring; the loop is hand-rolled and lives in `_compute_importance`."* The merging PR MUST add a one-paragraph note:

> **Implementation update (post-Option-A)**: the unconditional path (`strategy != "auto"` and no partitioner) now delegates to [`sklearn.inspection.permutation_importance`](https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html). The conditional path (subgroup-restricted permutation) remains hand-rolled because sklearn has no equivalent primitive. Public surface (FR-003.1, FR-003.4) is unchanged.

The spec edit lands **in the same PR** as the code change (Principle I: spec and code in lockstep).

## Option B — Close as `not_planned`

### Mechanical steps

```bash
# 1. Delete the local branch
git branch -D refactor/importance-sklearn

# 2. Confirm no remote ref
git ls-remote --heads origin refactor/importance-sklearn  # expect: empty
```

### Spec 003 update (mandatory part of closing)

Remove the dangling pointer from `specs/003-conditional-permutation-importance/spec.md` §"Out of scope / future work" — i.e. drop the paragraph cited under "Context" above and replace with:

> The unconditional PFI path is intentionally hand-rolled rather than delegated to `sklearn.inspection.permutation_importance`. The conditional path has no sklearn equivalent, and aligning the two implementations on a single hand-rolled engine is preferred for review-surface reasons. (Decision recorded in [issue #15](https://github.com/xeries-labs/xeries/issues/15).)

This edit lands as a small `docs(specs/003)` PR.

## Constitution Check

| Principle | Gate question | Verdict | Notes |
| --- | --- | --- | --- |
| I. Specs before code | Spec change lands with the code change? | **Pass** | Both options require an update to `specs/003-*/spec.md` in the same PR. |
| II. Agent-agnosticism | No agent-specific constructs introduced? | **Pass** | Pure library refactor. |
| III. Test-first for new contracts | At least one failing contract test? | **N/A** | Public surface unchanged; the existing cPFI tests are the contract. |
| IV. Typed public surface | `ty check src` and `ruff` pass? | **Pass** | Required for either option. |
| V. Reproducibility (bench only) | — | **N/A** | Home repo `xeries`. |
| Repo scope | Constitution repo-scope respected? | **Pass** | sklearn is already a runtime dep. |

## Definition of done

This issue closes when **either**:

- (Option A) the PR merges to `main` *and* the spec 003 update is in, *and* the local `refactor/importance-sklearn` branch is deleted (post-merge cleanup);  
   *or*
- (Option B) the spec 003 "Out of scope" paragraph is updated as above *and* the local branch is deleted, *and* this issue is closed with `state_reason=not_planned`.

## High-level tasks

- **T-15.1** Maintainer pick: Option A or Option B (this is the human-in-the-loop step).
- **T-15.2 (A)** Push branch, rebase, run cPFI suite + determinism check, verify lint/types.
- **T-15.3 (A)** Open PR including the FR-003.2 spec note; squash-merge after review.
- **T-15.2 (B)** Open a small `docs(specs/003)` PR removing the dangling pointer.
- **T-15.3 (B)** Close this issue with `state_reason=not_planned` once the PR merges.
- **T-15.4** Delete the local `refactor/importance-sklearn` branch.
